### ISSUE TITLE
Cannot install both Allegro 5.2 and libpng 1.6

### ISSUE THREAD

**Issue by fatcerberus** [liballeg/allegro5#904]
Allegro 5.2 seems to require libpng 1.2, attempting to install 1.6 causes parts of Allegro to be removed:
```
fatcerberus@pigcult-vm:~/src/minisphere$ sudo apt install libpng16-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
  gir1.2-gtk-2.0 liballegro-dialog5.2 liballegro-image5.2 liballegro-ttf5.2
  libatk1.0-dev libcairo-script-interpreter2 libexpat1-dev libharfbuzz-dev
  libharfbuzz-gobject0 libpixman-1-dev libwebp-dev libwebpdemux1
  libxcb-shm0-dev libxcomposite-dev libxi-dev libxml2-utils
  x11proto-composite-dev
Use 'sudo apt autoremove' to remove them.
The following additional packages will be installed:
  libpng16-16 libpng16-devtools
Recommended packages:
  libpng16-tools
The following packages will be REMOVED:
  liballegro-dialog5-dev liballegro-image5-dev liballegro-ttf5-dev
  libcairo2-dev libfontconfig1-dev libfreetype6-dev libgdk-pixbuf2.0-dev
  libgtk2.0-dev libpango1.0-dev libpng12-dev libxft-dev
The following NEW packages will be installed:
  libpng16-16 libpng16-dev libpng16-devtools
0 upgraded, 3 newly installed, 11 to remove and 0 not upgraded.
Need to get 425 kB of archives.
After this operation, 30.5 MB disk space will be freed.
Do you want to continue? [Y/n] 
```

After which the removed components cannot be reinstalled as the dependency cannot be met:
```
fatcerberus@pigcult-vm:~/src/minisphere$ sudo apt install liballegro-image5-dev
Reading package lists... Done
Building dependency tree       
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 liballegro-image5-dev : Depends: libpng12-dev but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```

Which subsequently causes builds to fail.  I'm not sure if anything can be done about this (the libpng versions are clearly mutually exclusive), but figured I'd bring it up.

**Comment by SiegeLord** [comment]
I think it's pretty typical for -dev packages to be mutually exclusive, even if the non-dev packages can be co-installed. In your command, for example, `libpng12-dev` is also removed when you install `libpng16-dev`, which chained it way down to Allegro. I think that's just the Debian life.

We probably could easily get Allegro to depend on `libpng16-dev` instead, but obviously we can't make it depend on either one, so people that for some reason want to use `libpng12-dev` will now run into the same issue you're describing.

**Comment by fatcerberus** [comment]
The main issue is that, from what I understand, libpng 1.2 and 1.6 are not source compatible.  Otherwise I would just target 1.2.x and be done with it.

I guess my only other alternative is to pull the actual libpng sources into my source tree, which I was trying to avoid, but what can you do...

**Comment by SiegeLord** [comment]
I'm all for getting Allegro to use libpng1.6 as an option, and default to it for Ubuntu. I guess, though, ultimately if there's no source compatibility, somebody screwed up in making their -dev packages exclusive (same thing is true for Allegro 4's and 5'th -dev packages, for no good reason).

**Comment by fatcerberus** [comment]
Well I fixed the issue on my end, anyway.  It turns out libpng 1.2 and 1.6 *are* source compatible as long as you don’t access the members of the `png_struct` directly (which was deprecated even in 1.2).

The reason for the separate packages is that actually hiding the struct members involved a breaking change to the ABI.  It’s still possible to write code that compiles with either version, thankfully.

**Comment by dos1** [comment]
Well, so if Allegro can be built with both, this issue is just a matter of package maintainer and the distro, isn't it? When the distro as a whole moves to libpng 1.6, its Allegro packages should get rebuilt.

**Comment by fatcerberus** [comment]
Yeah, that’s not the issue - the main issue is that on any distro where Allegro is built with libpng 1.2, if the Allegro-based app also uses libpng APIs directly then *it’s impossible to target 1.6* because the `-dev` packages for 1.2 and 1.6 are mutually exclusive.
